internal/shared/otlp/otlpmetric/selfobservability/selfobservability.go.tmpl

```diff
@@ -0,0 +1,184 @@
+	if err != nil {
+		em.enabled = false
+		return em
+	}
+
+	// Set up common attributes
+	em.attrs = []attribute.KeyValue{
+		semconv.OTelComponentTypeKey.String(componentType),
+		semconv.ServerAddress(serverAddress),
+		semconv.ServerPort(serverPort),
+	}
+
+	return em
+}
+
+// TrackExport tracks an export operation and returns a function to complete the tracking.
+// The returned function should be called when the export operation completes.
+func (em *ExporterMetrics) TrackExport(ctx context.Context, rm *metricdata.ResourceMetrics) func(error) {
+	if !em.enabled {
+		return func(error) {}
+	}
+
+	dataPointCount := countDataPoints(rm)
+	startTime := time.Now()
+
+	em.inflight.Add(ctx, dataPointCount, em.attrs...)
+
+	return func(err error) {
+		em.inflight.Add(ctx, -dataPointCount, em.attrs...)
+
+		duration := time.Since(startTime).Seconds()
+		attrs := em.attrs
+		if err != nil {
+			attrs = append(attrs, semconv.ErrorTypeOther)
+		}
+		em.duration.Inst().Record(ctx, duration, metric.WithAttributes(attrs...))
+
+		if err == nil {
+			em.exported.Add(ctx, dataPointCount, em.attrs...)
+		}
+	}
+}
+
+// countDataPoints counts the total number of data points in a ResourceMetrics.
+func countDataPoints(rm *metricdata.ResourceMetrics) int64 {
+	if rm == nil {
+		return 0
+	}
+
+	var total int64
+	for _, sm := range rm.ScopeMetrics {
+		for _, m := range sm.Metrics {
+			switch data := m.Data.(type) {
+			case metricdata.Gauge[int64]:
+				total += int64(len(data.DataPoints))
+			case metricdata.Gauge[float64]:
+				total += int64(len(data.DataPoints))
+			case metricdata.Sum[int64]:
+				total += int64(len(data.DataPoints))
+			case metricdata.Sum[float64]:
+				total += int64(len(data.DataPoints))
+			case metricdata.Histogram[int64]:
+				total += int64(len(data.DataPoints))
+			case metricdata.Histogram[float64]:
+				total += int64(len(data.DataPoints))
+			case metricdata.ExponentialHistogram[int64]:
+				total += int64(len(data.DataPoints))
+			case metricdata.ExponentialHistogram[float64]:
+				total += int64(len(data.DataPoints))
+			case metricdata.Summary:
+				total += int64(len(data.DataPoints))
+			}
+		}
+	}
+	return total
+}
+
+// parseEndpoint extracts server address and port from an endpoint URL.
+// Returns defaults if parsing fails.
+func ParseEndpoint(endpoint string, defaultPort int) (address string, port int) {
+	address = "localhost"
+	port = defaultPort
+
+	if endpoint == "" {
+		return
+	}
+
+	// Handle endpoint without scheme
+	if !strings.Contains(endpoint, "://") {
+		endpoint = "http://" + endpoint
+	}
+
+	u, err := url.Parse(endpoint)
+	if err != nil {
+		return
+	}
+
+	if u.Hostname() != "" {
+		address = u.Hostname()
+	}
+
+	if u.Port() != "" {
+		if p, err := strconv.Atoi(u.Port()); err == nil {
+			port = p
+		}
+	}
+
+	return
+}
+
+// isSelfObservabilityEnabled checks if self-observability is enabled via environment variable.
+func isSelfObservabilityEnabled() bool {
```

Comment:
I suggest referring to this example to implement the [`x.go`](https://github.com/open-telemetry/opentelemetry-go/pull/7027/files#diff-7a501024b2ec1eeabed37cf507f7dfdba83d79fec50c5474297325f61aab9baeR47-R63) file, as it adheres to certain normative requirements at the underlying level.

eg. https://github.com/open-telemetry/opentelemetry-specification/blob/62effed618589a0bec416a87e559c0a9d96289bb/specification/configuration/sdk-environment-variables.md#parsing-empty-value

> In addition, a consistent coding style has been maintained.